File: typescript/tests/boundPropagation.test.ts
// ===========================================================================
// BOUND PROPAGATION TESTS (FIX #9 Verification)
// ===========================================================================
// These tests verify that bounds propagate through arbitrary nesting levels
// to correctly stop repetitions before consuming delimiters.

import { testParse } from './testUtils.js';

describe('Bound Propagation Tests', () => {
  test('BP-01-direct-repetition', () => {
    // Baseline: Bound with direct Repetition child (was already working)
    const { ok, errorCount } = testParse('S <- "x"+ "end" ;', 'xxxxend');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('BP-02-through-ref', () => {
    // FIX #9: Bound propagates through Ref
    const grammar = `
      S <- A "end" ;
      A <- "x"+ ;
    `;
    const { ok, errorCount } = testParse(grammar, 'xxxxend');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('BP-03-through-nested-refs', () => {
    // FIX #9: Bound propagates through multiple Refs
    const grammar = `
      S <- A "end" ;
      A <- B ;
      B <- "x"+ ;
    `;
    const { ok, errorCount } = testParse(grammar, 'xxxxend');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('BP-04-through-first', () => {
    // FIX #9: Bound propagates through First alternatives
    const grammar = `
      S <- A "end" ;
      A <- "x"+ / "y"+ ;
    `;
    const { ok, errorCount } = testParse(grammar, 'xxxxend');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('BP-05-left-recursive-with-repetition', () => {
    // FIX #9: The EMERG-01 case - bound through LR + First + Seq + Repetition
    const grammar = `
      S <- E "end" ;
      E <- E "+" "n"+ / "n" ;
    `;
    const { ok, errorCount } = testParse(grammar, 'n+nnn+nnend');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('BP-06-with-recovery-inside-bounded-rep', () => {
    // FIX #9 + recovery: Bound propagates AND recovery works inside repetition
    const grammar = `
      S <- A "end" ;
      A <- "ab"+ ;
    `;
    const { ok, errorCount, skippedStrings } = testParse(grammar, 'abXabYabend');
    expect(ok).toBe(true);
    expect(errorCount).toBe(2);
    expect(skippedStrings).toContain('X');
    expect(skippedStrings).toContain('Y');
  });

  test('BP-07-multiple-bounds-nested-seq', () => {
    // Multiple bounds in nested Seq structures
    const grammar = `
      S <- A ";" B "end" ;
      A <- "x"+ ;
      B <- "y"+ ;
    `;
    const { ok, errorCount } = testParse(grammar, 'xxxx;yyyyend');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
    // A stops at ';', B stops at 'end'
  });

  test('BP-08-bound-vs-eof', () => {
    // Without explicit bound, should consume until EOF
    const { ok, errorCount } = testParse('S <- "x"+ ;', 'xxxx');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
    // No bound, so consumes all x's
  });

  test('BP-09-zeoormore-with-bound', () => {
    // Bound applies to ZeroOrMore too
    const { ok, errorCount } = testParse('S <- "x"* "end" ;', 'end');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('BP-10-complex-nesting', () => {
    // Deeply nested: Ref -> First -> Seq -> Ref -> Repetition
    const grammar = `
      S <- A "end" ;
      A <- "a" B / "fallback" ;
      B <- "x"+ ;
    `;
    const { ok, errorCount } = testParse(grammar, 'axxxxend');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  // ===========================================================================
  // Additional tests for transparent rule skipping and repetition bounds
  // ===========================================================================

  test('BPR-01-repetition-stops-at-sibling-terminal', () => {
    // OneOrMore should stop when sibling terminal can match
    const { ok, errorCount } = testParse('S <- "x"+ "Y" ;', 'xxY');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('BPR-02-repetition-stops-at-sibling-after-error', () => {
    // After recovering from error, repetition should still respect sibling
    const { ok, errorCount, skippedStrings } = testParse('S <- "x"+ "Y" ;', 'xZxY');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).toContain('Z');
    expect(skippedStrings).not.toContain('Y');
  });

  test('BPR-03-repetition-stops-at-optional-sibling', () => {
    // Repetition should stop when optional sibling can match
    const { ok, errorCount } = testParse('S <- "x"+ "!"? ;', 'xx!');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('BPR-04-repetition-with-error-stops-at-optional-sibling', () => {
    // After error recovery, repetition should stop at optional sibling
    const { ok, errorCount, skippedStrings } = testParse('S <- "x"+ "!"? ;', 'xZx!');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).not.toContain('!');
  });

  test('WBS-01-bracket-bound-through-whitespace', () => {
    // The "]" should be the effective bound, not WS
    const grammar = `
      S <- "[" WS Items WS "]" ;
      Items <- Item* ;
      Item <- "x" ;
      ~WS <- [ ]* ;
    `;
    const { ok, errorCount } = testParse(grammar, '[xx]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('WBS-03-bracket-bound-with-error-before-close', () => {
    // Error recovery should stop at "]", not consume it
    const grammar = `
      S <- "[" WS Items WS "]" ;
      Items <- Item* ;
      Item <- "x" ;
      ~WS <- [ ]* ;
    `;
    const { ok, errorCount, skippedStrings } = testParse(grammar, '[xZx]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).not.toContain(']');
  });

  test('ALS-01-simple-array-valid', () => {
    const grammar = `
      S <- "[" (V ("," V)*)? "]" ;
      V <- [0-9]+ ;
    `;
    const { ok, errorCount } = testParse(grammar, '[1,2,3]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('ALS-03-array-with-error-in-middle', () => {
    // Error between values should be recovered, ] should still match
    const grammar = `
      S <- "[" WS (V (WS "," WS V)*)? WS "]" ;
      V <- [0-9]+ ;
      ~WS <- [ ]* ;
    `;
    const { ok, errorCount, skippedStrings } = testParse(grammar, '[1,2X,3]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).toContain('X');
    expect(skippedStrings).not.toContain(']');
  });

  test('ALS-06-nested-arrays', () => {
    // Nested arrays - inner ] should not be consumed by outer repetition
    const grammar = `
      S <- "[" WS (V (WS "," WS V)*)? WS "]" ;
      V <- S / [0-9]+ ;
      ~WS <- [ ]* ;
    `;
    const { ok, errorCount } = testParse(grammar, '[[1,2],3]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('RDR-01-ref-to-whitespace-skipped', () => {
    // Transparent WS should be skipped, non-transparent RBRACKET should be bound
    const grammar = `
      S <- "[" WS Items WS RBRACKET ;
      Items <- Item* ;
      Item <- "x" ;
      ~WS <- [ ]* ;
      RBRACKET <- "]" ;
    `;
    const { ok, errorCount, skippedStrings } = testParse(grammar, '[xZx]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).not.toContain(']');
  });

  test('RSB-01-no-trailing-skip-past-delimiter', () => {
    // Trailing garbage recovery should stop at delimiter
    const grammar = `
      S <- "[" Items "]" ;
      Items <- "x"+ ;
    `;
    const { ok, skippedStrings } = testParse(grammar, '[xZ]');
    expect(ok).toBe(true);
    expect(skippedStrings).not.toContain(']');
  });

  test('CTR-01-charset-repetition-no-recovery', () => {
    // CharSet repetition should NOT try to skip over non-matching content
    const grammar = `
      S <- [a-z]+ "!" ;
    `;
    const { ok, errorCount } = testParse(grammar, 'abXcd!');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
  });

  test('CTR-03-complex-subclause-does-recovery', () => {
    // Repetition of complex clauses (Seq) SHOULD do recovery
    const grammar = `
      S <- ("ab")+ "!" ;
    `;
    const { ok, errorCount, skippedStrings } = testParse(grammar, 'abXab!');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).not.toContain('!');
  });

  // ===========================================================================
  // Additional tests to match Dart test coverage
  // ===========================================================================

  test('BPR-05-nested-repetition-respects-outer-bound', () => {
    const grammar = `
      S <- "[" Items "]" ;
      Items <- Item* ;
      Item <- "x" ;
    `;
    const { ok, errorCount } = testParse(grammar, '[xx]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('BPR-06-nested-repetition-with-error-respects-bound', () => {
    const grammar = `
      S <- "[" Items "]" ;
      Items <- Item* ;
      Item <- "x" ;
    `;
    const { ok, errorCount, skippedStrings } = testParse(grammar, '[xZx]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).not.toContain(']');
  });

  test('WBS-02-bracket-bound-with-actual-whitespace', () => {
    const grammar = `
      S <- "[" WS Items WS "]" ;
      Items <- (WS Item)* ;
      Item <- "x" ;
      ~WS <- [ ]* ;
    `;
    const { ok, errorCount } = testParse(grammar, '[ x x ]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('WBS-04-brace-bound-through-whitespace', () => {
    const grammar = `
      S <- "{" WS Items WS "}" ;
      Items <- Item* ;
      Item <- "x" ;
      ~WS <- [ ]* ;
    `;
    const { ok, errorCount, skippedStrings } = testParse(grammar, '{xZx}');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).not.toContain('}');
  });

  test('WBS-05-multiple-whitespace-layers', () => {
    const grammar = `
      S <- "[" WS1 WS2 Items WS1 WS2 "]" ;
      Items <- Item* ;
      Item <- "x" ;
      ~WS1 <- [ ]* ;
      ~WS2 <- [\\t]* ;
    `;
    const { ok, errorCount, skippedStrings } = testParse(grammar, '[xZx]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).not.toContain(']');
  });

  test('ALS-02-array-with-whitespace-valid', () => {
    const grammar = `
      S <- "[" WS (V (WS "," WS V)*)? WS "]" ;
      V <- [0-9]+ ;
      ~WS <- [ ]* ;
    `;
    const { ok, errorCount } = testParse(grammar, '[ 1 , 2 , 3 ]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('RDR-02-multiple-refs-whitespace-all-skipped', () => {
    const grammar = `
      S <- "[" SP TAB Items SP TAB RBRACKET ;
      Items <- Item* ;
      Item <- "x" ;
      ~SP <- [ ]* ;
      ~TAB <- [\\t]* ;
      RBRACKET <- "]" ;
    `;
    const { ok, errorCount, skippedStrings } = testParse(grammar, '[xZx]');
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).not.toContain(']');
  });

  test('RSB-02-recovery-scan-stops-at-delimiter', () => {
    const grammar = `
      S <- "[" Items "]" ;
      Items <- ("ab")+ ;
    `;
    const { ok, skippedStrings } = testParse(grammar, '[abZ]');
    expect(ok).toBe(true);
    expect(skippedStrings).not.toContain(']');
  });

  test('RSB-03-multiple-delimiters-nested', () => {
    const grammar = `
      S <- "[" Inner "]" ;
      Inner <- "{" Items "}" ;
      Items <- "x"+ ;
    `;
    const { ok, skippedStrings } = testParse(grammar, '[{xZ}]');
    expect(ok).toBe(true);
    expect(skippedStrings).not.toContain('}');
    expect(skippedStrings).not.toContain(']');
  });
});

File: java/src/test/java/com/squirrelparser/BoundPropagationTest.java
// ===========================================================================
// BOUND PROPAGATION TESTS (FIX #9 Verification)
// ===========================================================================
// These tests verify that bounds propagate through arbitrary nesting levels
// to correctly stop repetitions before consuming delimiters.

package com.squirrelparser;

import static com.squirrelparser.TestUtils.testParse;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class BoundPropagationTest {

    @Test
    void bp01_directRepetition() {
        // Baseline: Bound with direct Repetition child (was already working)
        var result = testParse("S <- \"x\"+ \"end\" ;", "xxxxend");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have 0 errors");
    }

    @Test
    void bp02_throughRef() {
        // FIX #9: Bound propagates through Ref
        String grammar = """
            S <- A "end" ;
            A <- "x"+ ;
            """;
        var result = testParse(grammar, "xxxxend");
        assertTrue(result.ok(), "should succeed (bound through Ref)");
        assertEquals(0, result.errorCount(), "should have 0 errors");
    }

    @Test
    void bp03_throughNestedRefs() {
        // FIX #9: Bound propagates through multiple Refs
        String grammar = """
            S <- A "end" ;
            A <- B ;
            B <- "x"+ ;
            """;
        var result = testParse(grammar, "xxxxend");
        assertTrue(result.ok(), "should succeed (bound through 2 Refs)");
        assertEquals(0, result.errorCount(), "should have 0 errors");
    }

    @Test
    void bp04_throughFirst() {
        // FIX #9: Bound propagates through First alternatives
        String grammar = """
            S <- A "end" ;
            A <- "x"+ / "y"+ ;
            """;
        var result = testParse(grammar, "xxxxend");
        assertTrue(result.ok(), "should succeed (bound through First)");
        assertEquals(0, result.errorCount(), "should have 0 errors");
    }

    @Test
    void bp05_leftRecursiveWithRepetition() {
        // FIX #9: The EMERG-01 case - bound through LR + First + Seq + Repetition
        String grammar = """
            S <- E "end" ;
            E <- E "+" "n"+ / "n" ;
            """;
        var result = testParse(grammar, "n+nnn+nnend");
        assertTrue(result.ok(), "should succeed (bound through LR)");
        assertEquals(0, result.errorCount(), "should have 0 errors");
    }

    @Test
    void bp06_withRecoveryInsideBoundedRep() {
        // FIX #9 + recovery: Bound propagates AND recovery works inside repetition
        String grammar = """
            S <- A "end" ;
            A <- "ab"+ ;
            """;
        var result = testParse(grammar, "abXabYabend");
        assertTrue(result.ok(), "should succeed");
        assertEquals(2, result.errorCount(), "should have 2 errors (X and Y)");
        assertTrue(result.skippedStrings().contains("X"), "should skip X");
        assertTrue(result.skippedStrings().contains("Y"), "should skip Y");
    }

    @Test
    void bp07_multipleBoundsNestedSeq() {
        // Multiple bounds in nested Seq structures
        String grammar = """
            S <- A ";" B "end" ;
            A <- "x"+ ;
            B <- "y"+ ;
            """;
        var result = testParse(grammar, "xxxx;yyyyend");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have 0 errors");
        // A stops at ';', B stops at 'end'
    }

    @Test
    void bp08_boundVsEof() {
        // Without explicit bound, should consume until EOF
        var result = testParse("S <- \"x\"+ ;", "xxxx");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have 0 errors");
        // No bound, so consumes all x's
    }

    @Test
    void bp09_zeroormoreWithBound() {
        // Bound applies to ZeroOrMore too
        var result = testParse("S <- \"x\"* \"end\" ;", "end");
        assertTrue(result.ok(), "should succeed (ZeroOrMore matches 0)");
        assertEquals(0, result.errorCount(), "should have 0 errors");
    }

    @Test
    void bp10_complexNesting() {
        // Deeply nested: Ref -> First -> Seq -> Ref -> Repetition
        String grammar = """
            S <- A "end" ;
            A <- "a" B / "fallback" ;
            B <- "x"+ ;
            """;
        var result = testParse(grammar, "axxxxend");
        assertTrue(result.ok(), "should succeed (bound through complex nesting)");
        assertEquals(0, result.errorCount(), "should have 0 errors");
    }

    // ===========================================================================
    // Additional tests for transparent rule skipping and repetition bounds
    // ===========================================================================

    @Test
    void bpr01_repetitionStopsAtSiblingTerminal() {
        // OneOrMore should stop when sibling terminal can match
        var result = testParse("S <- \"x\"+ \"Y\" ;", "xxY");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have no errors");
    }

    @Test
    void bpr02_repetitionStopsAtSiblingAfterError() {
        // After recovering from error, repetition should still respect sibling
        var result = testParse("S <- \"x\"+ \"Y\" ;", "xZxY");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error (Z)");
        assertTrue(result.skippedStrings().contains("Z"), "should skip Z");
        assertTrue(!result.skippedStrings().contains("Y"), "should NOT skip Y");
    }

    @Test
    void bpr03_repetitionStopsAtOptionalSibling() {
        // Repetition should stop when optional sibling can match
        var result = testParse("S <- \"x\"+ \"!\"? ;", "xx!");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have no errors");
    }

    @Test
    void bpr04_repetitionWithErrorStopsAtOptionalSibling() {
        // After error recovery, repetition should stop at optional sibling
        var result = testParse("S <- \"x\"+ \"!\"? ;", "xZx!");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error (Z)");
        assertTrue(!result.skippedStrings().contains("!"), "should NOT skip !");
    }

    @Test
    void wbs01_bracketBoundThroughWhitespace() {
        // The "]" should be the effective bound, not WS
        String grammar = """
            S <- "[" WS Items WS "]" ;
            Items <- Item* ;
            Item <- "x" ;
            ~WS <- [ ]* ;
            """;
        var result = testParse(grammar, "[xx]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have no errors");
    }

    @Test
    void wbs03_bracketBoundWithErrorBeforeClose() {
        // Error recovery should stop at "]", not consume it
        String grammar = """
            S <- "[" WS Items WS "]" ;
            Items <- Item* ;
            Item <- "x" ;
            ~WS <- [ ]* ;
            """;
        var result = testParse(grammar, "[xZx]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error (Z)");
        assertTrue(!result.skippedStrings().contains("]"), "should NOT skip ]");
    }

    @Test
    void als01_simpleArrayValid() {
        String grammar = """
            S <- "[" (V ("," V)*)? "]" ;
            V <- [0-9]+ ;
            """;
        var result = testParse(grammar, "[1,2,3]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have no errors");
    }

    @Test
    void als03_arrayWithErrorInMiddle() {
        // Error between values should be recovered, ] should still match
        String grammar = """
            S <- "[" WS (V (WS "," WS V)*)? WS "]" ;
            V <- [0-9]+ ;
            ~WS <- [ ]* ;
            """;
        var result = testParse(grammar, "[1,2X,3]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error (X)");
        assertTrue(result.skippedStrings().contains("X"), "should skip X");
        assertTrue(!result.skippedStrings().contains("]"), "should NOT skip ]");
    }

    @Test
    void als06_nestedArrays() {
        // Nested arrays - inner ] should not be consumed by outer repetition
        String grammar = """
            S <- "[" WS (V (WS "," WS V)*)? WS "]" ;
            V <- S / [0-9]+ ;
            ~WS <- [ ]* ;
            """;
        var result = testParse(grammar, "[[1,2],3]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have no errors");
    }

    @Test
    void rdr01_refToWhitespaceSkipped() {
        // Transparent WS should be skipped, non-transparent RBRACKET should be bound
        String grammar = """
            S <- "[" WS Items WS RBRACKET ;
            Items <- Item* ;
            Item <- "x" ;
            ~WS <- [ ]* ;
            RBRACKET <- "]" ;
            """;
        var result = testParse(grammar, "[xZx]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error (Z)");
        assertTrue(!result.skippedStrings().contains("]"), "should NOT skip ]");
    }

    @Test
    void rsb01_noTrailingSkipPastDelimiter() {
        // Trailing garbage recovery should stop at delimiter
        String grammar = """
            S <- "[" Items "]" ;
            Items <- "x"+ ;
            """;
        var result = testParse(grammar, "[xZ]");
        assertTrue(result.ok(), "should succeed");
        assertTrue(!result.skippedStrings().contains("]"), "should NOT skip ]");
    }

    @Test
    void ctr01_charsetRepetitionNoRecovery() {
        // CharSet repetition should NOT try to skip over non-matching content
        String grammar = """
            S <- [a-z]+ "!" ;
            """;
        var result = testParse(grammar, "abXcd!");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error");
    }

    @Test
    void ctr03_complexSubclauseDoesRecovery() {
        // Repetition of complex clauses (Seq) SHOULD do recovery
        String grammar = """
            S <- ("ab")+ "!" ;
            """;
        var result = testParse(grammar, "abXab!");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error (X)");
        assertTrue(!result.skippedStrings().contains("!"), "should NOT skip !");
    }

    // ===========================================================================
    // Additional tests to match Dart test coverage
    // ===========================================================================

    @Test
    void bpr05_nestedRepetitionRespectsOuterBound() {
        String grammar = """
            S <- "[" Items "]" ;
            Items <- Item* ;
            Item <- "x" ;
            """;
        var result = testParse(grammar, "[xx]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have no errors");
    }

    @Test
    void bpr06_nestedRepetitionWithErrorRespectsBound() {
        String grammar = """
            S <- "[" Items "]" ;
            Items <- Item* ;
            Item <- "x" ;
            """;
        var result = testParse(grammar, "[xZx]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error (Z)");
        assertTrue(!result.skippedStrings().contains("]"), "should NOT skip ]");
    }

    @Test
    void wbs02_bracketBoundWithActualWhitespace() {
        String grammar = """
            S <- "[" WS Items WS "]" ;
            Items <- (WS Item)* ;
            Item <- "x" ;
            ~WS <- [ ]* ;
            """;
        var result = testParse(grammar, "[ x x ]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have no errors");
    }

    @Test
    void wbs04_braceBoundThroughWhitespace() {
        String grammar = """
            S <- "{" WS Items WS "}" ;
            Items <- Item* ;
            Item <- "x" ;
            ~WS <- [ ]* ;
            """;
        var result = testParse(grammar, "{xZx}");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error (Z)");
        assertTrue(!result.skippedStrings().contains("}"), "should NOT skip }");
    }

    @Test
    void wbs05_multipleWhitespaceLayers() {
        String grammar = """
            S <- "[" WS1 WS2 Items WS1 WS2 "]" ;
            Items <- Item* ;
            Item <- "x" ;
            ~WS1 <- [ ]* ;
            ~WS2 <- [\t]* ;
            """;
        var result = testParse(grammar, "[xZx]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error (Z)");
        assertTrue(!result.skippedStrings().contains("]"), "should NOT skip ]");
    }

    @Test
    void als02_arrayWithWhitespaceValid() {
        String grammar = """
            S <- "[" WS (V (WS "," WS V)*)? WS "]" ;
            V <- [0-9]+ ;
            ~WS <- [ ]* ;
            """;
        var result = testParse(grammar, "[ 1 , 2 , 3 ]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have no errors");
    }

    @Test
    void rdr02_multipleRefsWhitespaceAllSkipped() {
        String grammar = """
            S <- "[" SP TAB Items SP TAB RBRACKET ;
            Items <- Item* ;
            Item <- "x" ;
            ~SP <- [ ]* ;
            ~TAB <- [\t]* ;
            RBRACKET <- "]" ;
            """;
        var result = testParse(grammar, "[xZx]");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error (Z)");
        assertTrue(!result.skippedStrings().contains("]"), "should NOT skip ]");
    }

    @Test
    void rsb02_recoveryScanStopsAtDelimiter() {
        String grammar = """
            S <- "[" Items "]" ;
            Items <- ("ab")+ ;
            """;
        var result = testParse(grammar, "[abZ]");
        assertTrue(result.ok(), "should succeed");
        assertTrue(!result.skippedStrings().contains("]"), "should NOT skip ]");
    }

    @Test
    void rsb03_multipleDelimitersNested() {
        String grammar = """
            S <- "[" Inner "]" ;
            Inner <- "{" Items "}" ;
            Items <- "x"+ ;
            """;
        var result = testParse(grammar, "[{xZ}]");
        assertTrue(result.ok(), "should succeed");
        assertTrue(!result.skippedStrings().contains("}"), "should NOT skip }");
        assertTrue(!result.skippedStrings().contains("]"), "should NOT skip ]");
    }
}
